echopop/spatial/transect.py

```diff
@@ -346,3 +363,130 @@ def transect_bearing(transect_data: pd.DataFrame):
 
     # Return the output
     return transect_direction.reset_index()
+
+
+def export_transect_layers(
+    transect_data: pd.DataFrame,
+    index_variable: Union[str, List[str]] = ["transect_num", "interval"],
+):
+    """
+    Calculate the mean depth and layer height for specific grouped data.
+    """
+
+    # Check that index variables exist
+    # ---- Convert to list, if needed
+    if isinstance(index_variable, str):
+        index_variable = list(index_variable)
+    elif not isinstance(index_variable, list):
+        raise TypeError(
+            f"The defined `region_filter` ({index_variable}) must be either a `str` or `list`."
+        )
+
+    # Check columns
+    # ---- Missing columns
+    missing_columns = set(index_variable) - set(transect_data.columns)
+    # ---- Raise error if needed
+    if missing_columns:
+        raise ValueError(
+            f"The following columns are missing from `transect_data`: {list(missing_columns)}"
+        )
+
+    # Calculate the maximum depth
+    if "max_depth" not in transect_data.columns:
+        raise ValueError("The column 'max_depth' is missing from `transect_data`.")
+    else:
+        # ---- Compute the bottom depth
+        transect_summary = (
+            transect_data.groupby(index_variable)["max_depth"].max().to_frame("bottom_depth")
+        )
+
+    # Calculate the mean layer depth and height
+    if "layer_depth_min" not in transect_data.columns:
+        raise ValueError("The column 'layer_depth_min' is missing from `transect_data`.")
+    elif "layer_depth_max" not in transect_data.columns:
+        raise ValueError("The column 'layer_depth_max' is missing from `transect_data`.")
+    else:
+        # ---- Mean layer depth
+        transect_summary["layer_mean_depth"] = (
+            transect_data.groupby(index_variable)
+            .agg(
+                mean_depth_min=("layer_depth_min", "min"), mean_depth_max=("layer_depth_max", "max")
+            )
+            .assign(mean_depth=lambda x: (x.mean_depth_min + x.mean_depth_max) / 2)["mean_depth"]
+        )
+        # ---- Mean layer height
+        transect_summary["layer_height"] = (
+            transect_data.groupby(index_variable)
+            .agg(
+                mean_depth_min=("layer_depth_min", "min"), mean_depth_max=("layer_depth_max", "max")
+            )
+            .assign(height=lambda x: x.mean_depth_max - x.mean_depth_min)["height"]
+        )
+
+    # Return the output
+    return transect_summary.reset_index()
+
+
+def export_transect_spacing(
```

Comment:
This loop below obviously gets the job done:
```
# Iterate through the transects to determine the maximum spacing
for i in range(len(transect_number)):
```
I am thinking that this could vectorized by groupby ops.
Since `latitude_threshold` is fixed, you can subset the dataframe first and then groupby based on `transect_num`, and compute mean, max, min from the groups. Then you could get the spacing with the right lag pretty easily (diff offset by 2), and then do the filtering (the if condition at the end) and assign them all.

Not sure how fast/slow the current ops are. If it is slow, this is worth considering.